### DISCUSSION TITLE
Relations: Generic class does not specify types

### DISCUSSION THREAD

**Discussion by joostdebruijn** [larastan/larastan#1960] [Ideas]
Hi!

I'm using Larastan v2.9.6 at Level 6 and I get messages like this on all relations:

```Method App\Models\UserAuthenticator::user() return type with generic class Illuminate\Database\Eloquent\Relations\BelongsTo does not specify its types: TRelatedModel, TChildModel```

I can solve this with defining doc blocks of course, but just for my understanding: is there no way this can be determined from the relation definition in the code?

**Comment by canvural** [comment]
Hi,

Yes, if you don't specify it Larastan still understands it. You can verify it by running something like `echo $userAuthenticator->user->email` and there should be no errors. But the error you mentioned will still be reported. You can ignore it with a regex if you want.

But I really recommend to add the docblocks. If you do not want to do it manually there is even a [Rector rule](https://github.com/driftingly/rector-laravel/blob/main/docs/rector_rules_overview.md#addgenericreturntypetorelationsrector) that can add them automatically.

**Comment by joostdebruijn** [comment]
Thanks for the quick response and the reference to the Rector rule, @canvural!
